### PR TITLE
Fix for gui where we didn't use the font resource correctly

### LINKED ISSUE #9653: Changing font resource for gui broken in 1.9.4
### Expected behaviour
Changing font for gui worked. (like in 1.9.3)

### Actual behaviour
Engine crashed.

https://github.com/user-attachments/assets/e765f9bc-c4c1-4d95-b8a3-36a104e395c7

```
INFO:CRASH: Successfully wrote Crashdump to file: C:\Users\ROG\AppData\Roaming\Defold/_crash
ERROR:CRASH: CALL STACK:

ERROR:CRASH:  0 0x7FF78E0CC6E0 dmCrash::GenerateCallstack D:\a\defold\defold\engine\crash\src\backtrace_win32.cpp:144
ERROR:CRASH:  1 0x7FF78E1BFC4C _seh_filter_exe minkernel\crts\ucrt\src\appcrt\misc\exception_filter.cpp:219
ERROR:CRASH:  2 0x7FF78E21F994 `__scrt_common_main_seh'::`1'::filt$0 D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:304
ERROR:CRASH:  3 0x7FF78E140EAC __C_specific_handler D:\a\_work\1\s\src\vctools\crt\vcruntime\src\eh\riscchandler.cpp:389
ERROR:CRASH:  4 0x7FFE4CEF5050 __chkstk <unknown>:0
ERROR:CRASH:  5 0x7FFE4CE6DDC0 RtlFindCharInUnicodeString <unknown>:0
ERROR:CRASH:  6 0x7FFE4CEF4140 KiUserExceptionDispatcher <unknown>:0
ERROR:CRASH:  7 0x19E87F3E430 <unknown symbol> <unknown>:0
ERROR:CRASH: 

INFO:CRASH: Successfully wrote Crashdump to file: C:\Users\ROG\AppData\Roaming\Defold/_crash
ERROR:CRASH: CALL STACK:

ERROR:CRASH:  0 0x19E87F3E430 <unknown symbol> <unknown>:0
ERROR:CRASH:  1 0x7FF78DEDA790 dmRender::Layout<dmRender::LayoutMetrics> D:\a\defold\defold\engine\render\src\render\font_renderer_private.h:90
ERROR:CRASH:  2 0x7FF78DEDF030 dmRender::DrawText D:\a\defold\defold\engine\render\src\render\font_renderer.cpp:622
ERROR:CRASH:  3 0x7FF78DD78A50 dmGameSystem::RenderTextNodes D:\a\defold\defold\engine\gamesys\src\gamesys\components\comp_gui.cpp:1146
ERROR:CRASH:  4 0x7FF78DD76100 dmGameSystem::RenderNodes D:\a\defold\defold\engine\gamesys\src\gamesys\components\comp_gui.cpp:2202
ERROR:CRASH:  5 0x7FF78E0C28D0 dmGui::RenderScene D:\a\defold\defold\engine\gui\src\gui.cpp:1464
ERROR:CRASH:  6 0x7FF78DD6FD10 dmGameSystem::CompGuiRender D:\a\defold\defold\engine\gamesys\src\gamesys\components\comp_gui.cpp:2519
ERROR:CRASH:  7 0x7FF78DD25190 dmGameObject::Render D:\a\defold\defold\engine\gameobject\src\gameobject\gameobject.cpp:2722
ERROR:CRASH:  8 0x7FF78DD041E0 dmEngine::StepFrame D:\a\defold\defold\engine\engine\src\engine.cpp:1740
ERROR:CRASH:  9 0x7FF78DD05030 dmEngineUpdate D:\a\defold\defold\engine\engine\src\engine.cpp:2275
ERROR:CRASH: 10 0x7FF78DD05260 dmEngine::RunLoop D:\a\defold\defold\engine\engine\src\engine_loop.cpp:83
ERROR:CRASH: 11 0x7FF78DCFD100 engine_main D:\a\defold\defold\engine\engine\src\engine_main.cpp:152
ERROR:CRASH: 12 0x7FF78E1352DC __scrt_common_main_seh D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288
ERROR:CRASH: 13 0x7FFE4C892560 BaseThreadInitThunk <unknown>:0
ERROR:CRASH: 14 0x7FFE4CEAAEE0 RtlUserThreadStart <unknown>:0
ERROR:CRASH: 
```

### Steps to reproduce
```lua
go.property("target_gui", msg.url())
go.property("font", resource.font("/font.font"))
go.property("font2", resource.font("/font2.font"))
function init(self)
	timer.delay(1, false, function()
		go.set(self.target_gui, "fonts", self.font2, { key = "font" })
	end)
end
```

[change_font.zip](https://github.com/user-attachments/files/17519608/change_font.zip)

<hr/>

<table>
<tr><td>Build time</td><td>2024-10-23T14:41:30.540039</td></tr>
<tr><td>Defold channel</td><td>editor-alpha</td></tr>
<tr><td>Defold editor sha</td><td>ed533d8f5deef70915d1cd4bf129998ade7bea80</td></tr>
<tr><td>Defold engine sha</td><td>a9277779d1c23b0712a5959314b9508d2e802efb</td></tr>
<tr><td>Defold version</td><td>1.9.4</td></tr>
<tr><td>GPU</td><td>Intel(R) RaptorLake-S Mobile Graphics Controller</td></tr>
<tr><td>GPU Driver</td><td>4.6.0 - Build 31.0.101.5186</td></tr>
<tr><td>Java version</td><td>17.0.5+8</td></tr>
<tr><td>OS arch</td><td>amd64</td></tr>
<tr><td>OS name</td><td>Windows 11</td></tr>
<tr><td>OS version</td><td>10.0</td></tr>
<table>

### PR DESCRIPTION
**Pull request by JCash**
Fixes https://github.com/defold/defold/issues/9653

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [ ] New and changed code follows the overall code style of existing code
	* [ ] Add comments where needed
* [ ] Documentation
	* [ ] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   2. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

### CODE REVIEW

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 1]
```diff
@@ -170,7 +170,7 @@ namespace dmGameSystem
         return node_material_res ? GetNodeMaterial(node_material_res) : gui_context->m_Material;
     }
 
-    static inline dmRender::HMaterial GetTextNodeMaterial(RenderGuiContext* gui_context, dmGui::HScene scene, dmGui::HNode node, dmRender::HFontMap font_map)
+    static inline dmRender::HMaterial GetTextNodeMaterial(RenderGuiContext* gui_context, dmGui::HScene scene, dmGui::HNode node, dmRender::HFont font_map)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:173]: HFont is the new public name in our sdk

File: engine/gamesys/src/gamesys/components/comp_gui.cpp [hunk 2]
```diff
@@ -2756,21 +2756,23 @@ namespace dmGameSystem
                 return dmGameObject::PROPERTY_RESULT_INVALID_KEY;
             }
             dmResource::HFactory factory = dmGameObject::GetFactory(params.m_Instance);
-            dmGameSystem::FontResource* font = 0;
-            dmGameObject::PropertyResult res = SetResourceProperty(factory, params.m_Value, FONT_EXT_HASH, (void**)&font);
+            dmGameSystem::FontResource* font_resource = 0;
+            dmGameObject::PropertyResult res = SetResourceProperty(factory, params.m_Value, FONT_EXT_HASH, (void**)&font_resource);
             if (res == dmGameObject::PROPERTY_RESULT_OK)
             {
+                dmRender::HFont font = dmGameSystem::ResFontGetHandle(font_resource);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_gui.cpp:2763]: The bug fix:
We recently updated from a HFont to a FontResource* as the actual resource pointer.
And I had missed this case. The gui system stores pointers to the HFont, whereas I accidentally passed in a FontResource*, making the types mismatched, and generated an invalid access.

File: engine/gamesys/src/gamesys/test/gui/goscript.script [hunk 3]
```diff
@@ -0,0 +1,5 @@
+function init(self)
+    timer.delay(.001, false, function()
+        go.set("/go#gui", "fonts", hash("/font/glyph_bank_test_1.fontc"), { key = "test_font" })
+    end)
+end
```
Comment (JCash) [inline engine/gamesys/src/gamesys/test/gui/goscript.script:5]: Same as in the repro case, triggering the code path in the comp_gui.cpp

**Review by Jhonnyg — DISMISSED** [review]
LGTM!

**Review by Jhonnyg — APPROVED** [review]

**Pull request merged**